sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/PruneHiveTablePartitions.scala

```diff
@@ -0,0 +1,140 @@
+ */
+private[sql] class PruneHiveTablePartitions(session: SparkSession)
+  extends Rule[LogicalPlan] with CastSupport {
+
+  override val conf = session.sessionState.conf
+
+  /**
+   * Extract the partition filters from the filters on the table.
+   */
+  private def extractPartitionPruningFilters(
+              filters: Seq[Expression],
+              relation: HiveTableRelation): Seq[Expression] = {
+    val normalizedFilters = filters.map { e =>
+      e transform {
+        case a: AttributeReference =>
+          a.withName(relation.output.find(_.semanticEquals(a)).get.name)
+      }
+    }
+    val partitionSet = AttributeSet(relation.partitionCols)
+    normalizedFilters.filter { predicate =>
+      !predicate.references.isEmpty && predicate.references.subsetOf(partitionSet)
+    }
+  }
+
+  /**
+   * Prune the hive table using filters on the partitions of the table.
+   */
+  private def prunePartitions(relation: HiveTableRelation, partitionFilters: Seq[Expression])
+  : Seq[CatalogTablePartition] = {
+    val partitions =
+    if (conf.metastorePartitionPruning) {
+      session.sessionState.catalog.listPartitionsByFilter(
+        relation.tableMeta.identifier, partitionFilters)
+    } else {
+      session.sessionState.catalog.listPartitions(relation.tableMeta.identifier)
+    }
+    val shouldKeep = partitionFilters.reduceLeftOption(And).map { filter =>
+      require(filter.dataType == BooleanType,
+        s"Data type of predicate $filter must be ${BooleanType.catalogString} rather than " +
+          s"${filter.dataType.catalogString}.")
+      BindReferences.bindReference(filter, relation.partitionCols)
+    }
+    if (shouldKeep.nonEmpty) {
+      partitions.filter{ partition =>
+        val hivePartition =
+          HiveClientImpl.toHivePartition(partition, HiveClientImpl.toHiveTable(relation.tableMeta))
+        val dataTypes = relation.partitionCols.map(_.dataType)
+        val castedValues = hivePartition.getValues.asScala.zip(dataTypes)
+          .map { case (value, dataType) => cast(Literal(value), dataType).eval(null) }
+        val row = InternalRow.fromSeq(castedValues)
+        shouldKeep.get.eval(row).asInstanceOf[Boolean]
+      }
+    } else {
+      partitions
+    }
+  }
+
+  /**
+   * Update the statistics of the table.
+   */
+  private def updateTableMeta(
+              tableMeta: CatalogTable,
+              prunedPartitions: Seq[CatalogTablePartition]): CatalogTable = {
+    val sizeInBytes = try {
+      prunedPartitions.map { partition =>
+        val rawDataSize = partition.parameters.get(StatsSetupConst.RAW_DATA_SIZE).map(_.toLong)
+        val totalSize = partition.parameters.get(StatsSetupConst.TOTAL_SIZE).map(_.toLong)
+        if (rawDataSize.isDefined && rawDataSize.get > 0) {
+          rawDataSize.get
+        } else if (totalSize.isDefined && totalSize.get > 0L) {
+          totalSize.get
+        } else {
+          CommandUtils.calculateLocationSize(
```

Comment:
It's too expensive to get size for file system for each partition. We should drop the size stats if we can't get it from the partition metadata.

Reply:
In PruneFileSourcePartitions, it calls "prunedFileIndex.sizeInBytes" which get size from file system for each leaf file (include files of all partitions). I think it's ok to get size from file system for each partition, but we need to limit the number of partitions, just as proposed in #27129 . 
Here, I changed it to drop the size stats if it can't be got from metadata, will update it after #27129 finished.
Does it make sense to you?

Reply:
SGTM